File: login-server/src/main/java/work/fking/pangya/login/packet/handler/CheckNicknamePacketHandler.java
package work.fking.pangya.login.packet.handler;

import io.netty.buffer.ByteBuf;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import work.fking.pangya.login.LoginServer;
import work.fking.pangya.login.Player;
import work.fking.pangya.login.net.ClientPacketHandler;
import work.fking.pangya.login.net.LoginState;
import work.fking.pangya.login.packet.outbound.CheckNicknameReplies;
import work.fking.pangya.networking.protocol.ProtocolUtils;

public class CheckNicknamePacketHandler implements ClientPacketHandler {

    private static final Logger LOGGER = LogManager.getLogger(CheckNicknamePacketHandler.class);

    @Override
    public void handle(LoginServer server, Player player, ByteBuf packet) {
        var nickname = ProtocolUtils.readPString(packet);
        var channel = player.channel();

        if (player.loginState() != LoginState.SELECTING_NICKNAME && player.loginState() != LoginState.SELECTED_NICKNAME) {
            LOGGER.warn("Unexpected login session state, got={}, expected=SELECTING_NICKNAME or SELECTED_NICKNAME", player.loginState());
            channel.disconnect();
            return;
        }
        channel.writeAndFlush(CheckNicknameReplies.available(nickname));
    }
}

File: login-server/src/main/kotlin/work/fking/pangya/login/packet/handler/CheckNicknamePacketHandler.kt
package work.fking.pangya.login.packet.handler

import io.netty.buffer.ByteBuf
import work.fking.pangya.login.LoginServer
import work.fking.pangya.login.Player
import work.fking.pangya.login.net.ClientPacketHandler
import work.fking.pangya.login.net.LoginState.SELECTED_NICKNAME
import work.fking.pangya.login.net.LoginState.SELECTING_NICKNAME
import work.fking.pangya.login.packet.outbound.CheckNicknameReplies
import work.fking.pangya.networking.protocol.readPString

class CheckNicknamePacketHandler : ClientPacketHandler {

    override fun handle(server: LoginServer, player: Player, packet: ByteBuf) {
        val nickname = packet.readPString()

        require(player.state == SELECTING_NICKNAME || player.state == SELECTED_NICKNAME) {
            "Unexpected state ${player.state}, expected SELECTING_NICKNAME or SELECTED_NICKNAME"
        }
        // TODO: validate if nickname is unique
        player.state = SELECTED_NICKNAME
        player.writeAndFlush(CheckNicknameReplies.available(nickname))
    }
}
